### ISSUE TITLE
HOW TO CREATE Dapp on THINKIUM

### ISSUE THREAD

**Issue by Cloudnapoleon** [ThinkiumGroup/wiki#11]
Step by step guide to building a decentralized application (dApp) on 
**THINKIUM**

If we were to build our voting application on the web, we’d run into a few problems:
The data on the database could be changed: it could be counted more than once, or removed entirely.

The source code on the webserver could also be changed at any time.
We don’t want to build our app on the web. We want to build it on the blockchain where anyone connected to the network can participate in the election. We want to ensure that their votes are counted and that they are only counted once. So let’s take a look at how that works.

Instead of having a network, a central server, and a database, the blockchain is a network and a database all in one. A blockchain is a peer-to-peer network of computers, called nodes, that share all the data and the code in the network. So, if you’re a device connected to the Thinkium blockchain, you are a node in the network, and you talk to all the other computer nodes in the network. You now have a copy of all the data and the code on the blockchain. There are no more central servers. Just a bunch of computers that talk to one another on the same network. 

Instead of a centralized database, all the transaction data that is shared across the nodes in the blockchain is contained in bundles of records called blocks, which are chained together to create the public ledger. This public ledger represents all the data in the blockchain. All the data in the public ledger is secured by cryptographic hashing and validated by a consensus algorithm. Nodes on the network participate to ensure that all copies of the data distributed across the network are the same. That’s one very important reason why we’re building our voting application on the blockchain because we want to ensure that our vote was counted and that it did not change.

Without much wasting of time LETS MOVE TO WHAT WE'LL BE  BUILDING

How to build a client-side application that will talk to our smart contract on the Thinkium blockchain. This client-side application will have a table of candidates that lists each candidate’s id, name, and vote count. It will have a form where we can cast a vote for our desired candidate. It also shows the account we’re connected to the blockchain under “your account”.

Installing Dependencies
In order to build our dApp, we need a few dependencies first.

Node Package Manager (NPM)
The first dependency we need is Node Package Manager, or NPM, which comes with Node.js. You can see if you have a node already installed by going to your terminal and typing:

$ node -v

Truffle Framework
The next dependency is the Truffle Framework, which allows us to build decentralized applications on the Ethereum blockchain which can also be used in Thinkium blockchain. It provides a suite of tools that allow us to write smart contracts with the Solidity programming language. It also enables us to test our smart contracts and deploy them to the blockchain. It also gives us a place to develop our client-side application. 

You can install Truffle with NPM in your command line like this:

$ npm install -g truffle

Ganache:
The next dependency is Ganache, a local in-memory blockchain. You can install Ganache by downloading it from this link : http://truffleframework.com/ganache . It will give us 10 external accounts with addresses on our local blockchain. Each account is preloaded with 100 fake ethers.

Metamask
The next dependency is the metamask chrome extension. https://chrome.google.com/webstore/detail/metamask/nkbihfbeogaeaoehlefnkodbefgpgknn?hl=en
In order to use the blockchain, we must connect to it (remember, I said the blockchain is a network). We’ll have to install a special browser extension in order to use the Thinkium blockchain. That’s where metamask comes in. We’ll be able to connect to our local Ethereum blockchain with our personal account and interact with our smart contract.

We’re going to be using the Metamask chrome extension for this tutorial, so you’ll also need to install the google chrome browser if you don’t have it already. To install Metamask, search for the Metamask Chrome plugin in the Google Chrome web store. Once you’ve installed it, be sure that it is checked in your list of extensions. You’ll see the fox icon in the top right-hand side of your Chrome browser when it’s installed. Reference the video walkthrough if you get stuck!

Step 1: Smoke Test
Now that we have our dependencies installed, let’s start building our dApp!
First, find where you downloaded Ganache, and open it. Now that Ganache has booted, you have a local blockchain running.

Now let’s create a project directory for our dApp in the command line like this:

$ mkdir election
$ cd election

Now that we’re inside our project, we can get up and running fast with a Truffle box. We’ll be using the Pet Shop box for this tutorial. From within your project directory, install the pet shop box from the command line like this:

$ truffle unbox pet-shop

Step 2: List Candidates
Now that everything is set up properly, let’s continue building out the smart contact by listing out the candidates that will run in the election. We need a way to store multiple candidates and store multiple attributes about each candidate. We want to keep track of a candidate’s id, name, and vote count. Here is how we will model the candidate:

contract Election {
    // Model a Candidate
    struct Candidate {
        uint id;
        string name;
        uint voteCount;
    }

// ...
}

We have modelled the candidate with a Solidity Struct. Solidity allows us to create our own structure types as we’ve done for our candidate here. We specified that this struct has an id of unsigned integer type, name of string type, and voteCount of unsigned integer type. Simply declaring this struct won’t actually give us a candidate. We need to instantiate it and assign it to a variable before we can write it to storage.

The next thing we need is a place to store the candidates. We need a place to store one of the structure types that we’ve just created. We can do this with a Solidity mapping. A mapping in Solidity is like an associative array or a hash, that associates key-value pairs. We can create this mapping like this:

contract Election {
    // Model a Candidate
    struct Candidate {
        uint id;
        string name;
        uint voteCount;
    }

/ Read/write Candidates
    mapping(uint => Candidate) public candidates;

// ...
}

In this case, the key to the mapping is an unsigned integer, and the value is a Candidate structure type that we just defined. This essentially gives us an id-based lookup for each candidate. Since this mapping is assigned to a state variable, we will write data to the blockchain anytime we assign new key-value pairs to it. Next, we set this mapping’s visibility to public in order to get a getter function, just like we did with the candidate name in the smoke test.

THE STEP BELOW IS FOR BEGINNERS
The steps below is also suitable for creating a Dapp on Thinkium blockchain

An environment setup is necessary to start the development process. Let us understand every step briefly.

1) Install Truffle and Ganache
Truffle is a testing and development framework that is used to build smart contracts for Ethereum which will also be suitable for THINKIUM. This tool provides an efficient way to build and deploy smart contracts to the blockchain.

DeFi dapps is also using another tool, Ganache. It allows creating local Ethereum blockchain to test smart contracts. This tool stimulates the blockchain network features and provides funds to the first ten accounts with 100 test Ether. Thus, it makes the deployment of smart contracts easy and free of cost.

2) Create An ERC20 Token
ERC20 is a fungible token that is used to stake on the smart contract. To develop an ERC20 token, you need to install an OpenZeppelin library. This library consists of implementation standards for ERC20 and ERC721.

3) Compile An ERC20 Token
To compile the smart contract, you must first check the solidity compiler version. You can check that by running the command:

Truffle Version

Solidity v0.5.16 is a default truffle version. If the token is written in the solidity version 0.6.2 if you run the command to compile your contracts you will get a compiler error. To specify which solidity compiler version to use, go to the file truffle-config.js and set it to your desired compile version.

4) Deploy ERC Token
Once the compilation is done, you can now deploy your token.

For this, open Ganache and select the option “Quickstart” to start a local Ethereum blockchain. To deploy your contract, run:truffle migrate.

To deploy the smart contract successfully, use the first address listed on the Ganache. To verify that, open the Ganache desktop app and can check the balance of Ether for the first account, it has been reduced due to the cost of Ether to deploy your smart contract.

To Sum Up
These are the required steps to build your DeFi dapp on Thinkium blockchain
